infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/EnvironmentVariableSecretApplier.java

```diff
@@ -0,0 +1,112 @@
+ * containers. Allows per-component control of secret applying in devfile.
+ */
+@Beta
+@Singleton
+public class EnvironmentVariableSecretApplier
+    extends KubernetesSecretApplier<KubernetesEnvironment> {
+
+  static final String ANNOTATION_ENV_NAME = ANNOTATION_PREFIX + "/" + "env-name";
+  static final String ANNOTATION_ENV_NAME_TEMPLATE = ANNOTATION_PREFIX + "/%s_" + "env-name";
+
+  @Override
+  public void applySecret(KubernetesEnvironment env, Secret secret) throws InfrastructureException {
+    boolean secretAutomount =
+        Boolean.parseBoolean(secret.getMetadata().getAnnotations().get(ANNOTATION_AUTOMOUNT));
+    for (PodData podData : env.getPodsData().values()) {
+      if (!podData.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      for (Container container : podData.getSpec().getContainers()) {
+        Optional<ComponentImpl> component = getComponent(env, container.getName());
+        if ((component.isPresent() && isComponentAutomountFalse(component.get()))
+            || (!secretAutomount
+                && !(component.isPresent() && isComponentAutomountTrue(component.get())))) {
+          continue;
+        }
+        for (Entry<String, String> secretDataEntry : secret.getData().entrySet()) {
+          final String mountEnvName = envName(secret, secretDataEntry.getKey());
+          container
+              .getEnv()
+              .add(
+                  new EnvVarBuilder()
+                      .withName(mountEnvName)
+                      .withValueFrom(
+                          new EnvVarSourceBuilder()
+                              .withSecretKeyRef(
+                                  new SecretKeySelectorBuilder()
+                                      .withName(secret.getMetadata().getName())
+                                      .withKey(secretDataEntry.getKey())
+                                      .build())
+                              .build())
+                      .build());
+        }
+      }
+    }
+  }
+
+  private String envName(Secret secret, String key) throws InfrastructureException {
+    String mountEnvName;
+    if (secret.getData().size() == 1) {
+      mountEnvName =
+          Stream.of(
+                  secret.getMetadata().getAnnotations().get(ANNOTATION_ENV_NAME),
+                  secret
+                      .getMetadata()
+                      .getAnnotations()
+                      .get(format(ANNOTATION_ENV_NAME_TEMPLATE, key)))
+              .filter(Objects::nonNull)
+              .findFirst()
+              .orElseThrow(
+                  () ->
+                      new InfrastructureException(
+                          format(
+                              "Unable to mount secret '%s': It is configured to be mount as a environment variable, but its was not specified. Please define the '%s' annotation on the secret to specify it.",
+                              secret.getMetadata().getName(), ANNOTATION_ENV_NAME)));
+    } else {
```

Comment:
does this apply also for `data < 1` ? I assume it's invalid state and its handled before calling this method.

Reply:
this method is called for each data entry. So in there is no entries, it will simply not called

Reply:
So it expects some state of arguments, but does not limit it programatically nor document it. Possible future caller of this method will need to dig in to discover this.